Python/Module6_Testing/Pytest.md

```diff
@@ -0,0 +1,750 @@
+tests\test_basic_numpy.py .                                              [100%]
+
+============================== 3 passed in 0.04s ==============================
+```
+
+
+This output indicates that three test-functions were found across two files and that all of the tests "passed"; i.e. the functions ran without raising any errors.
+The first two tests are located in `tests/test_basic_functions.py`; the two dots indicate that two functions were run, and the `[66%]` indicator simply denotes that the test-suite is 66% (two-thirds) complete.
+The proceeding reading comprehension problem will lead us to see what looks like for pytest to report a failing test.
+
+
+<div class="alert alert-info"> 
+
+**Reading Comprehension: Running a Test Suite**
+
+Temporarily add a new "broken" test to `tests/test_basic_functions.py`.
+The name that you give this test should adhere to pytest's simple rules for test-discovery.
+Design the test function so that is sure to fail when it is run.
+
+Rerun your test suite and compare its output to what you saw before - is it easy to identify which test failed and what caused it to fail?
+Make sure to remove this function from your test suite once you are finished answering this question. 
+
+</div>
+
+
+
+We can also direct pytest to run the tests in a specific .py file. E.g. executing:
+
+```shell
+pytest tests/test_basic_functions.py
+```
+
+will cue pytest to only run the tests in `test_basic_functions.py`.
+
+A key component to leveraging tests effectively is the ability to exercise ones tests repeatedly and rapidly with little manual overhead.
+Clearly, pytest is instrumental towards this end - this framework made the process of organizing and running our test suite exceedingly simple!
+That being said, there will certainly be occasions when we want to run a _specific_ test function.
+Suppose, for instance, that we are writing a new function, and repeatedly want to run one of our tests that is pointing to a bug in our work-in-progress.
+We can leverage pytest in conjunction with [an IDE](https://www.pythonlikeyoumeanit.com/Module1_GettingStartedWithPython/Getting_Started_With_IDEs_and_Notebooks.html) to run our tests in such incisive ways.
+
+
+### Utilizing pytest within an IDE
+
+Both [PyCharm and VSCode](https://www.pythonlikeyoumeanit.com/Module1_GettingStartedWithPython/Getting_Started_With_IDEs_and_Notebooks.html) can be configured to make keen use of pytest.
+The following images show a couple of the enhancements afforded to us by PyCharm; comparable features are available in VSCode.
+The IDEs will "discover" tests, and provide us with the ability to run individual tests.
+For example, in the following image, the green "play button" allows us to run `test_count_vowels_basic`.
+
+<!-- #raw raw_mimetype="text/html" -->
+<div style="text-align: center">
+<p>
+<img src="../_images/individual_test.PNG" alt="Running an individual test in PyCharm" width="600">
```

Comment:
`.PNG` -> `.png`, and same again below.  Windows doesn't care, but everything else does!